Src/Support/Google.Apis.Auth/OAuth2/LocalServerCodeReceiver.cs

```diff
@@ -493,65 +525,182 @@ internal class CallbackUriChooser
+                        // If they have the same amount of resets, then we prefer the one that has timed out
+                        // and we prefer loopback if both have timed out.
+                        _ when _loopbackIp.IsTimedOut => _loopbackIp,
+                        _ when _localhost.IsTimedOut => _localhost,
+                        // If they have the same amount of resets and none has timed out (they have failed), then we prefer loopback.
+                        _ => _loopbackIp
+                    };
+
+                    retriable.Reset();
+                    return retriable.Uri;
+                }
+
+                void InitUriStatisticsIfNeeded(ref UriStatistics statistics, string uri)
+                {
+                    if (statistics == null)
                     {
-                        // Ignore any errors in the constructor, they will re-occur later.
+                        statistics = new UriStatistics(uri, _timeout, _clock);
+
+                        // If possible, preemtively check that the uri works on this environment.
+                        // For instance, the loopback IP fails at least on Windows 7 and 8, for non-admin users.
+                        if (_listenerFailsFor(statistics.Uri))
+                        {
+                            statistics.Failed();
+                        }
                     }
+                }
+            }
+
+            public void ReportSuccess(string uri) => GetStatisticsFor(uri).Succeeded();
+
+            public void ReportFailure(string uri) => GetStatisticsFor(uri).Failed();
+
+            private UriStatistics GetStatisticsFor(string uri) =>
+                uri == CallbackUriTemplate127001 ? _loopbackIp :
+                uri == CallbackUriTemplateLocalhost ? _localhost :
+                throw new ArgumentOutOfRangeException(nameof(uri));
+
+            private static bool FailsHttpListener(string uri)
+            {
+#if NETSTANDARD1_3 || NETSTANDARD2_0
+                // No check required on NETStandard, it uses TcpListener which can only use IP adddresses, not DNS names.
+                return false;
+#elif NET45
+                try
+                {
+                    // This listener isn't used for anything except to check if it can listen on the given URI.
+                    // Hence it is disposed immediately.
+                    using var listener = new HttpListener();
+                    listener.Prefixes.Add(string.Format(uri, GetRandomUnusedPort()));
+                    listener.Start();
+                }
+                catch (HttpListenerException e) when (e.ErrorCode == 5) // 5: Access denied
+                {
+                    // Access denied for the given URI, report failure.
+                    return true;
+                }
+                catch
+                {
+                    // Ignore any errors in the constructor, they will re-occur later.
```

Comment:
nit: "in the constructor" no longer makes sense given the code refactoring. Maybe just "Ignore any errors here, they will re-occur later"?